Update the example in the guide (#371)
pygfx/pygfx: docs/guide.rst

pictures/videos of it).

.. note:: 
    The :ref:`full code example <full_example>` can be found below.

Let's look at a hello world example of how this works: A rotating cube. We begin
by defining an empty `Scene`::

    import pygfx as gfx

    scene = gfx.Scene()

Right now this scene is very desolate. There is no light, no objects, and
nothing that can look at those objects. Let's change this by adding some
`light` and a `camera`::

    # and god said ...
    scene.add(gfx.AmbientLight())

    camera = gfx.PerspectiveCamera(70, 16 / 9)
    camera.position.z = 400

Now there is light and a camera to perceive the light. To complete the setup
we also need to add an object to look at::

    geometry = gfx.box_geometry(200, 200, 200)
    material = gfx.MeshPhongMaterial(color=(1, 1, 0, 1))
    cube = gfx.Mesh(geometry, material)

    scene.add(cube)

Objects are slightly more complicated than lights or cameras. They have a
`geometry`, which controlls an object's form, and a `material`, which controls
an object's appearance (color, reflectiveness, etc).

Now we have all the necessary ingredients and it is time to take a look. To do

> so we need to create a `canvas` to draw what we see (here a on-screen window)

Reviewer: ```suggestion
so we need to create a `canvas` to draw what we see (here an on-screen window)
```

and a `renderer` that will look at the scene (through the camera) and draw what
it sees onto the canvas::

    from wgpu.gui.auto import WgpuCanvas, run

    renderer = gfx.renderers.WgpuRenderer(WgpuCanvas())
    renderer.render(scene, camera)

    run()

.. image:: _static/guide_static_cube.png

Nice, we rendered our first scene! While this technically completes the hello
world example, we can go one step further and add a bit of flare. We can add a
`draw_function` which allows us to add custom logic into the rendering process::

    def animate():
        # rotate the cube a bit
        rot = gfx.linalg.Quaternion().set_from_euler(gfx.linalg.Euler(0.005, 0.01))
        cube.rotation.multiply(rot)
        # update the canvas
        renderer.render(scene, camera)
        renderer.request_draw()

>     # make the cube pop

Reviewer: Pop may be confusion. Maybe something like:
```suggestion
    # Add a directional light to make the cube look better
```

Author: 😅

    scene.add(gfx.DirectionalLight())
    

>     # create a new window and new renderer

Reviewer: I think this may be confusing. My suggestion would be to only show the new code here (also not the `run()`. People can see a few lines down how things are put together.

Author: Fair. I was thinking of users that open an interpreter and copy past the bits into it while reading the guide. Those users will have to define a new canvas after closing it each time; hence the duplication.

Happy to update though, if you think this is more confusing than helpful.

Reviewer: Yeah, I figured you had an interactive session in mind, but that's not really clear here.

Actually, in an interactive session, you could add the light, define the animate function, call `request_draw` on it, and the original canvas would now show the cube rotating!

Author: @almarklein Oh, cool. In that case I think we should update to that and highlight it. Maybe we should even start with the renderer and a black screen and then make it come to life incrementally. What do you think?

Author: > Actually, in an interactive session, you could add the light, define the animate function, call request_draw on it, and the original canvas would now show the cube rotating!

I wanted to try this to see how to update the guide, but I didn't manage to make this work. `run()` is a blocking call and waits until the window is closed, so you can't modify the scene after calling it. 

At the same time, I don't get any render before calling `run()`. Instead, I get a light gray box and a friendly notification from Windows asking me if the program is stuck:

![image](https://user-images.githubusercontent.com/4402489/199826989-f6803734-88ca-47fd-91c4-e3a3e6ce5e58.png)

Author: To resolve this in an elegant way, I moved the `animate` extension into the `putting it all together` section. So now there is a step-by-step hello world example followed by a full example that adds a custom draw function. How does this look?

Reviewer: If there is a window, but no event loop that keeps it alive, it looks unresponsive to the OS.

The `run()`  function starts the event loop and is blocking by default. If run from an IDE with interactivity set up (with the Qt event loop if using a Qt widget, or with asyncio when using glfw) then `run` does not block and things are truely interactive.

However, this event loop stuff is a bit of a mess in Python, so we cannot expect the user to have it set up ...

I like how you solved it.

Author: > If run from an IDE with interactivity set up (with the Qt event loop if using a Qt widget, or with asyncio when using glfw) then run does not block and things are truely interactive.

Hm, I was running it in a normal python shell you'd get by calling the python executable. I guess you were referring to an ipython shell?

In any case, if you like the current solution then we can simply keep it like this and sort out the event loop when it becomes relevant the next time.

Reviewer: > I guess you were referring to an ipython shell?

Yes, for example, or any IDE that can do some event loop integration. But this is another topic indeed. Let's forget about interactive use for now and go with the current approach 👍

    renderer.request_draw(animate)
    run()

.. image:: _static/guide_rotating_cube.gif

.. _full_example:

Putting it together
-------------------

Combining everything introduced above into a single program, we obtain the
following::

    from wgpu.gui.auto import WgpuCanvas, run

    import pygfx as gfx

    scene = gfx.Scene()

    # create a camera to view the scene
    camera = gfx.PerspectiveCamera(70, 16 / 9)
    camera.position.z = 400

    # add some light so that the camera can see
    scene.add(gfx.AmbientLight())
    scene.add(gfx.DirectionalLight())

    # Populate a scene
    geometry = gfx.box_geometry(200, 200, 200)
    material = gfx.MeshPhongMaterial(color=(1, 1, 0, 1))
    cube = gfx.Mesh(geometry, material)
    scene.add(cube)
    
    # Create a canvas and a renderer
    renderer = gfx.renderers.WgpuRenderer(WgpuCanvas())

    def animate():
        # rotate the cube a bit
        rot = gfx.linalg.Quaternion().set_from_euler(gfx.linalg.Euler(0.005, 0.01))
        cube.rotation.multiply(rot)

        # redraw the scene
        renderer.render(scene, camera)
        renderer.request_draw()

    if __name__ == "__main__":

>         renderer.render(scene, camera)

Reviewer: ```suggestion
```

    scene = gfx.Scene()

Right now this scene is very desolate. There is no light, no objects, and
nothing that can look at those objects. Let's change this by adding some
`light` and a `camera`::

    # and god said ...
    scene.add(gfx.AmbientLight())

    camera = gfx.PerspectiveCamera(70, 16 / 9)
    camera.position.z = 400

Now there is light and a camera to perceive the light. To complete the setup
we also need to add an object to look at::

    geometry = gfx.box_geometry(200, 200, 200)
    material = gfx.MeshPhongMaterial(color=(1, 1, 0, 1))
    cube = gfx.Mesh(geometry, material)

    scene.add(cube)

Objects are slightly more complicated than lights or cameras. They have a
`geometry`, which controlls an object's form, and a `material`, which controls
an object's appearance (color, reflectiveness, etc).

Now we have all the necessary ingredients and it is time to take a look. To do
so we need to create a `canvas` to draw what we see (here an on-screen window)
and a `renderer` that will look at the scene (through the camera) and draw what
it sees onto the canvas::

    from wgpu.gui.auto import WgpuCanvas, run

    renderer = gfx.renderers.WgpuRenderer(WgpuCanvas())

>     renderer.render(scene, camera)

Reviewer: The `renderer.render()` call should be done in the "draw event".
```suggestion
    renderer.request_draw(lambda: renderer.render(scene, camera))
```

Lambda's are a bit ugly. Maybe it makes sense to create an `animate()` function here with just that one line, and explain the idea. Then down below the animate function is extended.

Author: oh? Did I get that right that you aren't supposed to call `renderer.render` from outside a `draw_function`?

Reviewer: Sorry for being a bit vague about this. I had to dig into the code a bit to get it straight :)  Let me try to write a piece of text that could be put in the guide (feel free to edit it to your style).

The `renderer.render(scene, camera)` call renders the scene, from the viewpoint of the camera, to the internal texture of the canvas. This does not put the image onto the screen yet; we need to request the canvas to perform a draw using `renderer.request_draw(animate)`. This makes the GUI schedule a draw event, in which it will call the given function, and then present the texture to the screen. Note that `renderer.request_draw()` can be called (without an argument) from hereon to schedule new draws.

Author: I've updated the guide to reflect this. @almarklein Please double check if I got the logic right and this makes sense.

communicate with the hardware. WGPU is itself based on Vulkan, Metal and DX12.

How to use pygfx?
-----------------

Pygfx is structured like many other rendering engines. We create a `scene` (a
worlds/scenarios to render) using three main ingredients: (1) `objects` and
their visual properties, (2) `light` sources, and (3) a `camera` that sees
things. Once we defined those three things, we can position them in our scene
and then use a `renderer` to look at what we have created (or take
pictures/videos of it).

.. note:: 
    The :ref:`full code example <full_example>` can be found below.

Let's look at a hello world example of how this works: A rotating cube. We begin
by defining an empty `Scene`::

    import pygfx as gfx

    scene = gfx.Scene()

Right now this scene is very desolate. There is no light, no objects, and
nothing that can look at those objects. Let's change this by adding some

> `light` and a `camera`::

Reviewer: The camera can be put as an object in the scene, but we don't do it here.
```suggestion
`light` and creating a `camera`::
```

    # and god said ...
    scene.add(gfx.AmbientLight())

    camera = gfx.PerspectiveCamera(70, 16 / 9)
    camera.position.z = 400

Now there is light and a camera to perceive the light. To complete the setup
we also need to add an object to look at::

    geometry = gfx.box_geometry(200, 200, 200)
    material = gfx.MeshPhongMaterial(color=(1, 1, 0, 1))
    cube = gfx.Mesh(geometry, material)

    scene.add(cube)

Objects are slightly more complicated than lights or cameras. They have a

> `geometry`, which controlls an object's form, and a `material`, which controls

Reviewer 2: ```suggestion
`geometry`, which controls an object's form, and a `material`, which controls
```

an object's appearance (color, reflectiveness, etc).

> Now we have all the necessary ingredients and it is time to take a look. This is
> done by a `renderer` and happens in a two step process known as a draw call:
> First, the renderer looks through the camera we created earlier, paints an image
> of what it sees, and stores it in a intermediate buffer. Second, the renderer
> takes this buffer and hands it over to a `canvas` which is responsible for
> displaying the image (here inside an on-screen window)::

Reviewer: Trying to make it a bit more precise:
```suggestion
Now we have all the necessary ingredients and it is time to take a look. This is
done by a `renderer`, which looks through the camera we created earlier, paints an image
of what it sees, and stores it in the internal buffer of the canvas. To display the result on
the screen, we need to request a draw. The canvas will schedule a good time to call our `draw_function`
and present the contents of its internal buffer to the screen when it returns.::
```

    from wgpu.gui.auto import WgpuCanvas, run

    canvas = WgpuCanvas()
    renderer = gfx.renderers.WgpuRenderer(canvas)

    def draw_function():
        # update the internal buffer
        renderer.render(scene, camera)

>         # Note: the handover to the canvas is implicit
> 

Reviewer: The internal buffer is already a responsibility of the canvas. Also at the low level (the canvas provides the texture to render to). So the handover is already inside `renderer.render()`.
```suggestion
```

    # schedule the draw call
    renderer.request_draw(draw_function)

    # run the application
    run()

.. image:: _static/guide_static_cube.png

Nice, we rendered our first scene! Next-up we can look at a full example that
takes this one step further and animates the cube.

.. _full_example:

Animations
----------

As promised in the previous section, here is a full example of how to use pygfx.

> It adds a little bit of flare to the hello world example by rotating the cube a

Reviewer 2: ```suggestion
It adds a little bit of flair to the hello world example by rotating the cube a
```

bit during the draw call. This allows us to create a simple animation::

    from wgpu.gui.auto import WgpuCanvas, run

    import pygfx as gfx

    scene = gfx.Scene()

    # create a camera to view the scene
    camera = gfx.PerspectiveCamera(70, 16 / 9)
    camera.position.z = 400

    # add some light so that the camera can see
    scene.add(gfx.AmbientLight())
    scene.add(gfx.DirectionalLight())

    # Populate the scene
    geometry = gfx.box_geometry(200, 200, 200)
    material = gfx.MeshPhongMaterial(color=(1, 1, 0, 1))
    cube = gfx.Mesh(geometry, material)
    scene.add(cube)
    
    # Create a canvas and a renderer
    canvas = WgpuCanvas()
    renderer = gfx.renderers.WgpuRenderer(canvas)

    def draw_function():
        # custom logic to rotate the cube
        rot = gfx.linalg.Quaternion().set_from_euler(
            gfx.linalg.Euler(0.005, 0.01)
        )
        cube.rotation.multiply(rot)

        # update the internal buffer
        renderer.render(scene, camera)

>         # Note: the handover to the canvas is implicit

Reviewer: ```suggestion
```

        # schedule the next draw call to show an animation
        # Note: without arguments it will use the previous draw_function

>         renderer.request_draw()

Reviewer: Worth noting (here, in the PR) that `renderer.request_draw()` is a wrapper for `canvas.request_draw()`. We use the renderer's version because then we don't need a reference to the canvas, leading to simpler code. So I don't think we need to mention it in the guide.

Author: Hm, so as a user of pygfx would I be expected to not interact with the canvas, or would I need to configure it somehow?

My angle is that, if I am expected to interact with a canvas, then I would wonder what the value of the `renderer.request_draw` function is. However, if I am not expected to directly interact with a canvas, then I would wonder why we need to introduce it as one of the first things in the guide. 

Following that thought, I would also wonder if it would make sense to just have a default value for the canvas in `gfx.renderers.WgpuRenderer()` (my suggestion is the `auto` canvas) that a user can overwrite if they need to. That way we only need to bring in a canvas.

Reviewer: Sorry if I only added confusion :)  I just wanted to mention it as "background". I don't think we should mention it in the guide, because in pygfx you typically don't need the canvas.

Author: > because in pygfx you typically don't need the canvas.

In that case I would vote for leaving the canvas out of the guide completely. Perhaps we can combine this with a feature PR that makes the auto canvas the default canvas for a renderer?

Reviewer: I would be against hiding it completely because it's still a useful concept. We could discuss this further, but let's do that in another issue/pr.

Author: This is now resolved by #378 and #380.

**Canvases**

The second main building block is the `Canvas`. A `Canvas` provides the surface
onto which the scene should be rendered and to use it you directly import it
from wgpu-py (on top of which pygfx is built). Wgpu-py has several canvases that
we can choose from, but for starters the most important ones is ``auto``, which
automatically selects an appropriate backend to talk to the GPU and the OS::

    # @almarklein: Can we provide a more meaningful example
    import pygfx as gfx
    from wgpu.gui.auto import WgpuCanvas

    canvas = WgpuCanvas()
    cube = gfx.Mesh(
        gfx.box_geometry(200, 200, 200),
        gfx.MeshPhongMaterial(color="#336699"),
    )

    if __name__ == "__main__":
        gfx.show(cube, canvas=canvas)

.. image:: _static/guide_hello_world.png

If you run the above, you will notice that it looks exactly the same as our
introductory example. Just as with a `Scene` above, `gfx.show` will create a
`Canvas` for you unless you provide it explicitly, and by default this uses the
``auto`` backend and hence we won't see any changes.

Why then would we explicitly provide a `Canvas`? Well, beyond the `Canvas`

> itself a backend also provides an event loop. This event loop allows us to

Reviewer: I think by far the most common use-case for providing a custom canvas is when someone builds a GUI, and wants gfx to render into a subwidget. If think should just say that here. It speaks more to the imagination, and now the user knows that PyGfx can indeed do that. The part about the event loop is true, but may be confusing to newcomers.

animate the scene or build a GUI by registering callbacks. To ensure that this
works correctly we need to use the same backend for both the event loop and the
`Canvas` and, for more complex scenarios, we may thus wish to create the
`Canvas` explicitly.

**Renderers**

The third and final main building block is a `Renderer`. A `Renderer` is like an
artist that brings all of the above together. It looks at the `Scene` through a
`Camera` and draws what it sees onto the surface provided by the `Canvas`. Like

> any good artist, a `Renderer` is never seen without its `Canvas`, so to create a

Reviewer: > Like any good artist, a `Renderer` is never seen without its `Canvas`

😆

`Renderer` we also need to create a `Canvas`::

    # @almarklein: Can we provide a more meaningful example
    import pygfx as gfx
    from wgpu.gui.auto import WgpuCanvas

    canvas = WgpuCanvas()
    renderer = gfx.renderers.WgpuRenderer(canvas)
    
    cube = gfx.Mesh(
        gfx.box_geometry(200, 200, 200),
        gfx.MeshPhongMaterial(color="#336699"),
    )

    if __name__ == "__main__":
        gfx.show(cube, renderer=renderer)

.. image:: _static/guide_hello_world.png

As you may expect by now, the output is the same as without the explicit
reference because `gfx.show` will create a renderer if you don't provide it. For
many applications this is perfectly fine; however, if we want to tackle more
advanced problems (e.g., control the exact process on how objects appear to
overlay each other) we may need to create it explicitly. For starters, it is
enough to know that it exists and what it does, so that we can come back to it
later when it becomes relevant.

(@almarklein: Do we have examples that customize the renderer that we could

> reference here?)

Author: @almarklein Hm, I was thinking about this as well but was wondering if this isn't too advanced a subject for a newcomer. The same for the blend mode. It's an excellent example, but I feel it would be better placed into a dedicated tutorial on blend modes and why they are important.

Re-reading the guide now, I think it could be fine to have the example as is. How do you think?

> reference here?)

Reviewer: Yeah, mentioning that the renderer allows advanced tuning of the rendering process, like you do now, is fine! 👍
